### DISCUSSION TITLE
Developers! How are you using Polyglot Notebooks or .NET Interactive

### DISCUSSION THREAD

**Discussion by KathleenDollard** [dotnet/interactive#2925] [General]
If you're a developer, let us know how you are using Polyglot Notebooks or .NET Interactive. 

Please include the language/kernel you're using, what your overall goal is, and how well it works/any problems your encounter.

**Comment by IntegerMan** [comment]
I use Polyglot Notebooks for:

- Mermaid diagrams
- ML .NET model training
- C# snippets and examples
- Teaching others

I'm looking into REST tonight and then variable sharing and the HTML / JS and then SQL and KQL for variety. I'll be getting a lot deeper into this area for my conference presentation on Polyglot and Jupyter on May 5th at Stir Trek.

I absolutely love Polyglot Notebooks and want to see it reach a point where I can load data, transform it, use it to train a model, and use it to visualize data as well.

**Comment by SpaceShot** [comment]
I have used them to integrate C# and F# with Markdown to use as a teaching aid.  I also used the SQL query as a way to have several queries that built on each other in different cells as an alternative to having SSMS open with different SQL queries, some in comment blocks because they're "less safe" than others, etc.  I imagine we all have had SSMS open with SQL just hanging out.  I used notebooks to try and be a little more formal and make something to share with my team.

I recently worked somewhere where we were getting into notebook in Databricks, and we had a heavy investment in C#.  There was real interest in having C# or .NET interactive work in that setting, but I couldn't figure it out so instead the developers just started using Python and made the switch.  I still am not sure if this is possible or maybe it's not really all that related to dotnet interactive.

Thanks for starting this discussion, I haven't tried using the Mermaid support but I am going to try using that next for a future presentation.  I think being able to deliver all of this in a GitHub repo that an attendee could clone and check out later has the potential to be really useful and a nice training aid after the fact.

**Comment by xg5-simon** [comment]
Loving it for PowerShell. Especially instructional material.
Do want to look at KQL and Kusto more in Polygot but find the documentation a little lacking there.

**Comment by SmartInteractiveGroup** [comment]
Playing with this extension. I was trying to create a notebook with PowerShell kernel that would create resources in Azure but I think the AzurePowerShell kernel is missing. But I am very exited with that particular extension. Best Regards.

**Comment by dfinke** [comment]
Just published a repo for OpenAI/Azure OpenAI PowerShell Samples: https://github.com/dfinke/openai-powershell-samples.
All Polyglot Interactive Notebooks.

**Reply by HowardvanRooijen** [reply]
Saw this and it's great! Well done!

**Reply by dfinke** [reply]
Thank you, much appreciated!  I don't seem to get these notifications.

**Comment by mika76** [comment]
I use it for loads of stuff. Definitely lots of conversions. I generate classes and interfaces from databases, convert data, do file translations, just like everything... It's amazing. Mostly use C# kernel but have played with F# and python.

**Comment by akrotov** [comment]
I would use it in polyglot notebooks. But when they got some initial features, like C# intellicence. Without intellicence it is like to write code in notepad.exe. Possible, but unneeded time killer. Besides intellisense it lucks support of any normal charting library. Currently Plotly.NET port is barely usable from C#. Some types of chart is impossible to create (vertically stacked from example) from C#. Maybe it is good for F#, haven't tried, but with C# it is highly ugly, non intuitive API. 
My typical task is machine learning (yes I am one of this strange .NET devs who tries to practice ML on .net stack). 
For now I am working around this limitation using paid LinqPad and opening of plotlye.net charts (yes it is ugly, but some types of charts are non convenient but usable, and I just didn't find any alternative with relative possibilities).

For now I can't image real use case for .NET Interactive. And not because of .NET interactive itself - the library is very good and promising. Because there is no polyglot-compatible notebooks with simple intellicence support.

**Reply by jonsequitur** [reply]
Intellisense is supported. If you're not seeing it, could you open an issue so we can dig into why it's not working for you? It might be something specific to your environment.

**Reply by akrotov** [reply]
Hm. Wait a second. Am I miss fix of intellisense? I was so hardly waiting for this feature to return back and was checking time to time. Last time when I tried was like 6 months ago and it was in more less state, like just checked today:

![image](https://github.com/dotnet/interactive/assets/3247982/527782c8-bde2-4917-a730-283de5311ae1)
![image](https://github.com/dotnet/interactive/assets/3247982/c99e61c7-14c3-468b-8446-19f692ad1451)

So technically this is an intellicence ofcoruse. But it just lists words from current text file that's all. No methods list no event standard namespaces. Is it expected state or I am facing an issue? Actually almost nothing changed, since I last time tried it (~6 months ago), only this strange "string" hint appeared after var. But maybe it is some 3rd party extension updated, not sure (haven't used VScode for also almost 6 months). 6 months ago I've tried literally everything. I was googling for the solutions for a while, then I've found some open issue in polyglot extension github where it was stated that intellisence for C# was "limited" by some performance issues until they are fixed. So I've decided that it I am just experiences this limited functionality. Tried a bit to find last version when it was worked to build it manually, but then just switched to LinqPad back because lack of time for this and waiting for the fix since that time) 

So can you please give a hint, is it an issue on screenshot or it is expected?

**Reply by jonsequitur** [reply]
It's hard to tell from your screen shot but it looks like the Polyglot Notebooks extension needs to be reloaded, which could cause this problem. 

<img width="371" alt="image" src="https://github.com/dotnet/interactive/assets/547415/18de2e84-119d-4dd6-8027-3f3e61dddafc">

You can refer to our troubleshooting suggestions here: https://github.com/dotnet/interactive/blob/main/docs/FAQ.md#troubleshooting

**Comment by Expecho** [comment]
I would like to use it for investigating issues and bugs but I am still waiting for KQL support against a log analytics workspace or application insight resource. There are 2 open issues that are asking for this feature, see [this](https://github.com/dotnet/interactive/issues/2849) and [this](https://github.com/dotnet/interactive/issues/1800). So far there does not seem to be much activity.

**Reply by jonsequitur** [reply]
Both of these KQL fixes are still on the roadmap but they depend on a more complex set of improvements to authentication (#992) and, as a result, to input handling and magic command parsing (#3194, #3323).

**Reply by Expecho** [reply]
Well, at least it is promising that they are still on the roadmap to be solved,

**Reply by Expecho** [reply]
@jonsequitur any status update you are able to share? Can I find this issue on any public roadmap somewhere?

**Comment by fforay** [comment]
I use it mainly to teach C# to undergraduate students.
I've done some Notebooks, for C#, SQL, EntityFramework, ML.NET, ...

Recently, I've also created a new Language Kernel in order to provide some Notebooks to beginners in X# Language (www.xsharp.eu)

**Comment by IntegerMan** [comment]
I'm starting a new series of videos over the summer teaching the C#
language to beginners and I happen to be using Polyglot Notebooks for a lot
of that process because you get nice inline display of values and short
feedback loops from writing code to seeing results.

Of course, I'm also using it to show the community what's possible with
data science in .NET and have an upcoming book and online video course on
this topic.

- Matt Eland, Microsoft MVP in AI
[image:
https://mvp.microsoft.com/en-us/PublicProfile/5005207?fullName=Matt%20Elandhttps://mvp.microsoft.com/en-us/PublicProfile/5005207?fullName=Matt%20Eland]
<https://mvp.microsoft.com/en-us/PublicProfile/5005207?fullName=Matt%20Elandhttps://mvp.microsoft.com/en-us/PublicProfile/5005207?fullName=Matt%20Eland>

**Comment by charabaruk** [comment]
I'm using notebooks to experiment on and shape development of various projects, generally for working out how things should function and work together, and if I'm on the right track with regard to API design in those projects.

**Comment by edward-encoord** [comment]
Besides all the cool stuff mentioned before, I plan to write one-off scripts to test my classes, which contain logics for different mathematical models. Many times, I need to quickly create a test case with multiple objects, to help communicate with clients/colleagues. Writing these code interactively frees me from compilation errors. I image storing a bunch of notebooks outside the codebase, so these small scripts (or hopefully ideas) don't get lost.

I do have one difficulty, so want to post it here. Some of my classes are internal, as we don't want the clients to rely on their API. They are however essential, as other classes depend on them. The difficulty is that these internal classes are not accessible in .NET interactive. This is a typical situation for a test project ([handled by `InternalsVisibleTo`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute?view=net-8.0)), but I [can't control the assembly of an interactive session](https://github.com/dotnet/roslyn/issues/12450).

**Reply by edward-encoord** [reply]
I mean the big advantages over a test project are:

- No need to exclude out-of-date files from the test project. Sometimes they contain useful information.
- Easier to change dependent packages.
- Run directly without compilation.
- Can export to a HTML file, maybe for presentation.

**Comment by a-t-k** [comment]
Danke euch für das klasse Tool. Bitte macht weiter. Das ist genau das Richtige um das Wissen in einer guten Qualität weiter zu geben.
Die letzte von mir erstellte und durchgeführte Präsentation über "Chanels" hatte sehr gute Resonanz. Es ist aber nicht nur "Präsentieren" sondern noch mehr auf die Fragen der Teilnehme anzugehen und dies gleich umsetzen. 

In letzter Zeit, werden von mir Polyglots geschrieben die unsere Projekte mit Beispielen und deren Nutzung dokumentieren. Echt klasse Tool, dafür.

Wenn weitere Featuren deren Reife erreichen, sollte das auch als Test-Tool eingesetzt werden. WebApi, DB, REPL. Mit einschließendem Reporting. 

Danke noch mal

**Reply by GeorgeS2019** [reply]
Copilot translation: 

Thank you for this great tool. Please keep it up. It is exactly the right thing to pass on knowledge in good quality. The last presentation I created and conducted on "Channels" received very positive feedback. But it's not just about "presenting," it's also about addressing the participants' questions and implementing them immediately.

Recently, I have been writing polyglots that document our projects with examples and their usage. Really great tool for that.

When more features reach maturity, it should also be used as a testing tool. Web API, DB, REPL. Including reporting.

Thanks again.

**Comment by jacobrastad** [comment]
I would very much like to use Notebooks for one-off / rarely used scripts in our large project. There are times when we need to do some sort of manipulation of data where you need the libs and models used by the large project. But we don't want to add it as a part of the large application or create some other project for it. Today we have several LinqPad variants of this where a Notebook would make more sense. And be much more usable with the power of markdown together with runnable code.

However I have not yet found a way to debug a C# Polyglot Notebook (as you can with Python & Jupyter) which makes it quite hard to use.  So really hoping for this in the near future.

Relevant issues for adding debug support:

- https://github.com/dotnet/interactive/issues/3446
- https://github.com/dotnet/interactive/issues/2099

**Comment by nczsl** [comment]
Jupyter for the. NET kernel is a very important tool and an important field for C # scripting applications. Python has achieved the status of the world's number one application language in this field. And personally, I feel that this field is also very useful and important because it does not compile, responds quickly, and provides an interactive code experience that other C # fields cannot achieve; There are currently many 'C # script' engines based on 'rosyln', but almost all of them are mediocre because their code execution experience is still far inferior to that of traditional scripting languages such as Python, with noticeable lag. However, compared to these C # script engines, VSCode Jupyter has a better interactive experience and is still smooth to use, similar to Python's Jupyter in many aspects. But it has many long-standing unresolved issues, such as the locking problem of the directory where DLLs are located; And the issue of Jupyter not correctly reflecting updates after DLL updates. For example, when you re execute a new DLL referenced by # r, sometimes it does not actually update, etc; There are also some apparent vulnerabilities, such as the possibility of 'vscode jupyter' crashing when executed multiple times in a short period of time. Anyway, I feel like there are still quite a few bugs. But as someone who loves scripting languages very much, I currently have no other choice if I want to use C # for script development. In addition, 'vscode jpyter' relies on the 'dotnet interactive' tool, but they all update slowly and still do not support. NET 9; I feel very uncomfortable. Why is the development so slow in these very important fields? The quality is not very good yet, and many bugs can affect the user experience, making it difficult to establish the overall ecosystem. Think about why Python has such a powerful ecosystem? I think it may be related to their good script user experience. So I hope these issues can be gradually improved in the future. Also, personally I won't install redundant features, so I won't install tools like 'polyglot notebooks' because I mainly use C # scripts and care about their quality, bugs, and performance issues. I don't want to use irrelevant or complicated things.

**Reply by GeorgeS2019** [reply]
Reorganize the points :-)

---
### Introduction

**Jupyter for the .NET Kernel:**
- A very important tool and field for C# scripting applications.
- Python has achieved the status of the world's number one application language in this field.
- This field is useful and important because it does not compile, responds quickly, and provides an interactive code experience that other C# fields cannot achieve.

### Issues with Current C# Script Engines

**Roslyn-based Engines:**
- Many C# script engines are based on Roslyn, but they are mediocre.
- Their code execution experience is inferior to traditional scripting languages like Python, with noticeable lag.

### Advantages of VSCode Jupyter

**Interactive Experience:**
- VSCode Jupyter offers a better interactive experience compared to other C# script engines.
- It is smooth to use and similar to Python's Jupyter in many aspects.

### Existing Issues with VSCode Jupyter

**Long-Standing Unresolved Issues:**
- Locking problem of the directory where DLLs are located.
- Jupyter not correctly reflecting updates after DLL updates.
  - For example, re-executing a new DLL referenced by #r sometimes does not actually update.
- Possibility of VSCode Jupyter crashing when executed multiple times in a short period of time.

### Personal Experience and Preferences

**Preference for Scripting Languages:**
- Despite the bugs, there are currently no better choices for C# script development.
- Reliance on the 'dotnet interactive' tool, which updates slowly and still does not support .NET 9.
- Discomfort with the slow development and existing bugs affecting the user experience.
- Comparisons with Python’s powerful ecosystem and good script user experience.

### Hope for Improvement

**Future Improvements:**
- Hopes for gradual improvements in these issues.
- Preference for not installing redundant features like 'polyglot notebooks.'
- Focus on the quality, bugs, and performance issues of C# scripts without using irrelevant or complicated tools.

---

**Comment by GeorgeS2019** [comment]
# Support for **UDF** when using Dotnet Spark.NET

The module needed for supporting UDF for Spark.NET in Interactive needs an upgrade, since there is now significant update to Spark.NET since 2025

https://github.com/dotnet/interactive/issues/3797

![image](https://github.com/user-attachments/assets/1b1e3090-ccb2-4e55-bf61-c658ae59c92a)

**Comment by mattsull-psu** [comment]
I'm just starting with polyglot, but as a SEIT/SDET, I use notebooks all the time because I can mix markdown with the results of queries, and have a record of what was the result of an execution of code at a particular moment in time. That way I can easily document the results at each step for a process with workflows. 

It *could* be an incredibly powerful tool if I am able to, for example, query a table and compare the results to a dataframe that I generated by using Playwright to go out and pull a table off a website.  Or using KQL to see what is in ADX, and then see if there are any differences with data from a SQL table.  Using Notebooks also makes it easy to repeat test processes in the future. 

Right now, I am running into trouble when I'm setting up the python kernel.  Why isn't that one of the default languages??  It may be that .NET developers overwhelmingly choose F# or C#, but Python and R are much more prevalent for data analysis and ML, it would seem to me Python should be available out of the box.

**Reply by jonsequitur** [reply]
A goal of #2867 is to make Python more obvious and discoverable. 

The reason why it's not available out of the box is that Python isn't assumed to be available on your machine, and when it is available, there are often multiple installations, virtual environments, etc. and discovering them at runtime is very slow, so there needs to be a way for you to configure some defaults.

**Comment by chaosdinosaur** [comment]
I've used Polyglot notebooks in the past with Powershell for demoing code or troubleshooting and running commands. The biggest blocker that I run into with fully using them instead of a Jupyter notebook is that Polyglot notebooks don't support bash language. If that was available to use it would open up an additional world of use for many of the projects I am working on where Bash and Python are being utilized at the same time.

The notebooks are really great for demoing outputs, especially if you have long running queries because it saves the outputs and you can check in for some documentation what a user should expect.

**Comment by darrellp** [comment]
I've been doing my Advent of Code problems in them.  After all - I'd hate for any of my genius in these matters to go undocumented! :)  Seriously though, I'm retired so I code mainly for fun and I like to think carefully about what I'm doing and record those thoughts in something more flexible and more interesting than in code comments.
I use mostly C# but I'm really learning about how powerful PowerShell is after all these years so it's great as a learning tool for PowerShell also.  I've loved Literate Programming ever since Knuth came up with the idea so this is right down my alley.  I do wish that it was more text based like Knuth's so that I could put one line of a method over here, another over there, disregard the order and have it all come together in the end.  I realize that I should be careful what I wish for and that this sort of thing could be more klunky in the end but man - I love the way Knuth's books read.
Main drawback IMHO is the lack of debugger/breakpoints.  If things get too ugly I transfer code into Linqpad where I do have those things and work with it there and then transfer it back after I've got things worked out but it would be much nicer if I didn't have to do that.
